src/main/scala/workflow/graph/NodeOptimizationRule.scala

```diff
@@ -0,0 +1,200 @@
+package workflow.graph
+
+import org.apache.spark.rdd.RDD
+import workflow.WorkflowUtils
+
+/**
+ * A modified graph executor used by the NodeOptimizationRule, to keep track
```

Comment:
Yes, this class is for walking through the graph and collecting samples for the node-level optimization. I'll rename it to `SampleCollector` and mention that in the docs. The actual process of walking through the DAG is somewhat changed, which is why I didn't make it extend GraphExecutor.

Notable differences : It doesn't optimize, doesn't keep track of prefixes, must sample all RDDs and later unpersist them, and must keep track of numPerPartition at each node.